pytorch_lightning/utilities/cli.py

```diff
@@ -325,41 +341,40 @@ def link_optimizers_and_lr_schedulers(self) -> None:
                 add_class_path = _add_class_path_generator(class_type)
                 self.parser.link_arguments(key, link_to, compute_fn=add_class_path)
 
-    def parse_arguments(self) -> None:
-        """Parses command line arguments and stores it in self.config"""
-        self.config = self.parser.parse_args()
+    def parse_arguments(self, parser: LightningArgumentParser) -> None:
+        """Parses command line arguments and stores it in ``self.config``."""
+        self.config = parser.parse_args()
 
     def before_instantiate_classes(self) -> None:
-        """Implement to run some code before instantiating the classes"""
+        """Implement to run some code before instantiating the classes."""
 
     def instantiate_classes(self) -> None:
-        """Instantiates the classes using settings from self.config"""
+        """Instantiates the classes and sets their attributes."""
         self.config_init = self.parser.instantiate_classes(self.config)
         self.datamodule = self.config_init.get("data")
         self.model = self.config_init["model"]
-        self.instantiate_trainer()
-
-    def instantiate_trainer(self) -> None:
-        """Instantiates the trainer using self.config_init['trainer']"""
-        if self.config_init["trainer"].get("callbacks") is None:
-            self.config_init["trainer"]["callbacks"] = []
         callbacks = [self.config_init[c] for c in self.parser.callback_keys]
-        self.config_init["trainer"]["callbacks"].extend(callbacks)
+        self.trainer = self.instantiate_trainer(self.config_init["trainer"], callbacks)
+
+    def instantiate_trainer(self, config: Dict[str, Any], callbacks: List[Callback]) -> Trainer:
```

Comment:
callbacks appearing here as an argument seems a bit random. what will happen in the future, will loggers etc. be added here?

Reply:
Depends.

If you look at the code, you'll see that this list of callbacks is added to the list of callbacks in the trainer config itself.

```python
config["callbacks"] = config["callbacks"] or []
config["callbacks"].extend(callbacks)
```
This is because support for default callbacks defined outside of the trainer config was added in #7964.

With these changes, I'm trying to avoid passing the entire config dictionaries to the `instantiate_trainer` method, but instead just pass the trainer config.

Since these "default" callbacks are not inside the trainer config, they also need to be passed as an input to the method.

So, if in the future we implement the same for loggers, yes we would need to pass them too.